src/imperative/cached_op.cc

```diff
@@ -0,0 +1,463 @@
+      } else {
+        ret.emplace_back(NodeEntry{p, k++, 0});
+      }
+    }
+  } else {
+    for (uint32_t i = 0; i < num_inputs(); ++i) ret.emplace_back(NodeEntry{p, i, 0});
+  }
+  return ret;
+}
+
+nnvm::Graph Imperative::CachedOp::GetForwardGraph(
+    const bool recording, const std::vector<NDArray*>& inputs) {
+  using namespace nnvm;
+  using namespace imperative;
+  std::lock_guard<std::mutex> lock(mutex_);
+  CHECK_EQ(inputs.size(), num_inputs());
+  nnvm::Graph& g = fwd_graph_;
+
+  ShapeVector shape_inputs;
+  DTypeVector dtype_inputs;
+  StorageTypeVector storage_type_inputs;
+  shape_inputs.reserve(inputs.size());
+  dtype_inputs.reserve(inputs.size());
+  storage_type_inputs.reserve(inputs.size());
+  for (uint32_t i = 0; i < inputs.size(); ++i) {
+    shape_inputs.emplace_back(inputs[i]->shape());
+    dtype_inputs.emplace_back(inputs[i]->dtype());
+    storage_type_inputs.emplace_back(inputs[i]->storage_type());
+  }
+
+  bool match = true;
+  match &= CheckAndInferShape(&g, std::move(shape_inputs), true);
+  match &= CheckAndInferType(&g, std::move(dtype_inputs), true);
+  match &= CheckAndInferStorageType(&g, inputs[0]->ctx(),
+                                    std::move(storage_type_inputs), true);
+
+  if (!match) {
+    g.attrs.erase("forward_mem_plan");
+    g.attrs.erase("full_mem_plan");
+  } else if (g.attrs.count(recording ? "full_mem_plan" : "forward_mem_plan")) {
+    return g;
+  }
+
+  const auto& idx = g.indexed_graph();
+
+  StorageVector storage(idx.num_node_entries(), exec::kBadStorageID);
+  for (const auto i : idx.input_nodes()) storage[idx.entry_id(i, 0)] = exec::kExternalStorageID;
+
+  auto mem_plan = PlanMemory(
+      &g, std::move(storage), g.GetAttr<std::vector<uint32_t> >(
+          recording ? "full_ref_count" : "forward_ref_count"));
+  g.attrs[recording ? "full_mem_plan" : "forward_mem_plan"] =
+      std::make_shared<dmlc::any>(std::move(mem_plan));
+
+  return g;
+}
+
+nnvm::Graph Imperative::CachedOp::GetBackwardGraph(
+    const OpStatePtr& op_state,
+    const std::vector<OpReqType>& reqs,
+    const std::vector<NDArray*>& inputs) {
+  using namespace nnvm;
+  using namespace imperative;
+  std::lock_guard<std::mutex> lock(mutex_);
+  nnvm::Graph& g = full_graph_;
+  auto& state = op_state.get_state<CachedOpState>();
+  bool req_match = true;
+  for (size_t i = 0; i < reqs.size(); ++i) {
+    if (curr_grad_req_[i] != (reqs[i] != kNullOp)) {
+      curr_grad_req_[i] = reqs[i] != kNullOp;
+      req_match = false;
+    }
+  }
+  if (!req_match) {
+    g = nnvm::Graph();
+    g.outputs = fwd_graph_.outputs;
+    for (size_t i = 0; i < grad_graph_.outputs.size(); ++i) {
+      if (curr_grad_req_[i]) g.outputs.emplace_back(grad_graph_.outputs[i]);
+    }
+  }
+
+  const auto& idx = g.indexed_graph();
+  size_t num_forward_nodes = fwd_graph_.indexed_graph().num_nodes();
+  size_t num_forward_entries = fwd_graph_.indexed_graph().num_node_entries();
```

Comment:
Is `num_forward_nodes` and `num_forward_entries ` from fwd_graph correct? The cached op takes the output of other ops and treats it as its input, which is not part of the fwd_graph inputs.